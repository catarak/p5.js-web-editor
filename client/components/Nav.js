import React, { PropTypes } from 'react';
import { Link } from 'react-router';

function Nav(props) {
  return (
    <nav className="nav" role="navigation" title="main-navigation">
      <ul className="nav__items-left" title="project-menu">
        <li className="nav__item">
          <a
            className="nav__new"
            onClick={props.newProject}
          >
            New
          </a>
        </li>
        <li className="nav__item">
          <a
            className="nav__save"
            onClick={props.saveProject}
          >
            Save
          </a>
        </li>
        {(() => { // eslint-disable-line
          if (props.project.id) {
            return (
              <li className="nav__item">
                <a className="nav__clone" onClick={props.cloneProject}>
                  Duplicate
                </a>
              </li>
            );
          }
        })()}
        {(() => { // eslint-disable-line
          if (props.project.id) {
            return (
              <li className="nav__item">
                <a className="nav__export" onClick={props.exportProjectAsZip}>
                  Download
                </a>
              </li>
            );
          }
        })()}
        {(() => { // eslint-disable-line
          if (props.project.id) {
            return (
              <li className="nav__item">
                <Link to={`/full/${props.project.id}`} target="_blank">
                  Fullscreen
                </Link>
              </li>
            );
          }
        })()}
        {(() => { // eslint-disable-line
          if (props.user.authenticated) {
            return (
              <li className="nav__item">
                <p className="nav__open">
                  <Link to="/sketches">
                    Open
                  </Link>
                </p>
              </li>
            );
          }
        })()}
        <li className="nav__item">
          <p className="nav__about">
            <Link to="/about">
              About
            </Link>
          </p>
        </li>
      </ul>
      <ul className="nav__items-right" title="user-menu">
        <li className="nav__item">
          {props.user.authenticated && <p>Hello, {props.user.username}!</p>}
          {!props.user.authenticated && <p><Link to="/login">Login</Link> or <Link to="/signup">Sign Up</Link></p>}
        </li>
      </ul>
    </nav>
  );
}

Nav.propTypes = {
  newProject: PropTypes.func.isRequired,
  saveProject: PropTypes.func.isRequired,
  exportProjectAsZip: PropTypes.func.isRequired,
  cloneProject: PropTypes.func.isRequired,
  user: PropTypes.shape({
    authenticated: PropTypes.bool.isRequired,
    username: PropTypes.string
  }).isRequired,
  project: PropTypes.shape({
    id: PropTypes.string
  })
};

export default Nav;
