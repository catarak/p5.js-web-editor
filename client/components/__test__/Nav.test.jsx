
import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import { NavComponent } from './../Nav';

describe('Nav', () => {
  const props = {
    newProject: jest.fn(),
    saveProject: jest.fn(),
    autosaveProject: jest.fn(),
    exportProjectAsZip: jest.fn(),
    cloneProject: jest.fn(),
    user: {
      authenticated: true,
      username: 'new-user',
      id: 'new-user'
    },
    project: {
      id: 'new-project',
      owner: {
        id: 'new-user'
      }
    },
    logoutUser: jest.fn(),
    newFile: jest.fn(),
    newFolder: jest.fn(),
    showShareModal: jest.fn(),
    showErrorModal: jest.fn(),
    unsavedChanges: false,
    warnIfUnsavedChanges: jest.fn(),
    showKeyboardShortcutModal: jest.fn(),
    cmController: {
      tidyCode: jest.fn(),
      showFind: jest.fn(),
      findNext: jest.fn(),
      findPrev: jest.fn()
    },
    startSketch: jest.fn(),
    stopSketch: jest.fn(),
    setAllAccessibleOutput: jest.fn()
  };
  const getWrapper = () => shallow(<NavComponent {...props} />);

  test('it renders main navigation', () => {
    const nav = getWrapper();
    expect(nav.exists('.nav')).toEqual(true);
  });

  it('renders correctly', () => {
    const tree = renderer
      .create(<NavComponent {...props} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
