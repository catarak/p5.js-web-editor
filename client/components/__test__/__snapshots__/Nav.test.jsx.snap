// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nav renders correctly 1`] = `
<header>
  <nav
    className="nav"
    title="main-navigation"
  >
    <ul
      className="nav__items-left"
    >
      <li
        className="nav__item-logo"
      >
        <test-file-stub
          aria-label="p5.js Logo"
          className="svg__logo"
          focusable="false"
          role="img"
        />
      </li>
      <li
        className="nav__item"
      >
        <button
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onMouseOver={[Function]}
        >
          <span
            className="nav__item-header"
          >
            File
          </span>
          <test-file-stub
            aria-hidden="true"
            className="nav__item-header-triangle"
            focusable="false"
          />
        </button>
        <ul
          className="nav__dropdown"
        >
          <li
            className="nav__dropdown-item"
          >
            <button
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
            >
              New
            </button>
          </li>
          <li
            className="nav__dropdown-item"
          >
            <button
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
            >
              Duplicate
            </button>
          </li>
          <li
            className="nav__dropdown-item"
          >
            <button
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
            >
              Share
            </button>
          </li>
          <li
            className="nav__dropdown-item"
          >
            <button
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
            >
              Download
            </button>
          </li>
          <li
            className="nav__dropdown-item"
          >
            <a
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              style={Object {}}
            >
              Open
            </a>
          </li>
        </ul>
      </li>
      <li
        className="nav__item"
      >
        <button
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onMouseOver={[Function]}
        >
          <span
            className="nav__item-header"
          >
            Edit
          </span>
          <test-file-stub
            aria-hidden="true"
            className="nav__item-header-triangle"
            focusable="false"
          />
        </button>
        <ul
          className="nav__dropdown"
        >
          <li
            className="nav__dropdown-item"
          >
            <button
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
            >
              Tidy Code
              <span
                className="nav__keyboard-shortcut"
              >
                ⇧
                +Tab
              </span>
            </button>
          </li>
          <li
            className="nav__dropdown-item"
          >
            <button
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
            >
              Find
              <span
                className="nav__keyboard-shortcut"
              >
                ⌃
                +F
              </span>
            </button>
          </li>
          <li
            className="nav__dropdown-item"
          >
            <button
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
            >
              Find Next
              <span
                className="nav__keyboard-shortcut"
              >
                ⌃
                +G
              </span>
            </button>
          </li>
          <li
            className="nav__dropdown-item"
          >
            <button
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
            >
              Find Previous
              <span
                className="nav__keyboard-shortcut"
              >
                ⇧
                +
                ⌃
                +G
              </span>
            </button>
          </li>
        </ul>
      </li>
      <li
        className="nav__item"
      >
        <button
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onMouseOver={[Function]}
        >
          <span
            className="nav__item-header"
          >
            Sketch
          </span>
          <test-file-stub
            aria-hidden="true"
            className="nav__item-header-triangle"
            focusable="false"
          />
        </button>
        <ul
          className="nav__dropdown"
        >
          <li
            className="nav__dropdown-item"
          >
            <button
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
            >
              Add File
            </button>
          </li>
          <li
            className="nav__dropdown-item"
          >
            <button
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
            >
              Add Folder
            </button>
          </li>
          <li
            className="nav__dropdown-item"
          >
            <button
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
            >
              Run
              <span
                className="nav__keyboard-shortcut"
              >
                ⌃
                +Enter
              </span>
            </button>
          </li>
          <li
            className="nav__dropdown-item"
          >
            <button
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
            >
              Stop
              <span
                className="nav__keyboard-shortcut"
              >
                ⇧
                +
                ⌃
                +Enter
              </span>
            </button>
          </li>
        </ul>
      </li>
      <li
        className="nav__item"
      >
        <button
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onMouseOver={[Function]}
        >
          <span
            className="nav__item-header"
          >
            Help
          </span>
          <test-file-stub
            aria-hidden="true"
            className="nav__item-header-triangle"
            focusable="false"
          />
        </button>
        <ul
          className="nav__dropdown"
        >
          <li
            className="nav__dropdown-item"
          >
            <button
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
            >
              Keyboard Shortcuts
            </button>
          </li>
          <li
            className="nav__dropdown-item"
          >
            <a
              href="https://p5js.org/reference/"
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              rel="noopener noreferrer"
              target="_blank"
            >
              Reference
            </a>
          </li>
          <li
            className="nav__dropdown-item"
          >
            <a
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              style={Object {}}
            >
              About
            </a>
          </li>
        </ul>
      </li>
    </ul>
  </nav>
</header>
`;
